cmd/promtool/debug.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/google/pprof/profile"
+)
+
+type debugWriterConfig struct {
+	server         string
+	tarballName    string
+	pathToFileName map[string]string
+	postProcess    func(b []byte) (*bytes.Buffer, error)
+}
+
+type debugWriter struct {
+	archiver
+	httpClient
+	requestToFile map[*http.Request]string
+	postProcess   func(b []byte) (*bytes.Buffer, error)
+}
+
+func newDebugWriter(cfg debugWriterConfig) *debugWriter {
+	client, err := newHTTPClient(httpClientConfig{serverURL: cfg.server})
+	if err != nil {
+		panic(err)
+	}
+	archiver, err := newArchiver(archiverConfig{archiveName: cfg.tarballName})
+	if err != nil {
+		panic(err)
+	}
+	reqs := make(map[*http.Request]string)
+	for path, filename := range cfg.pathToFileName {
+		req, err := http.NewRequest(http.MethodGet, client.urlJoin(path), nil)
+		if err != nil {
+			panic(err)
+		}
+		reqs[req] = filename
+	}
+	return &debugWriter{
+		archiver,
+		client,
+		reqs,
+		cfg.postProcess,
+	}
+}
+
+func (w *debugWriter) Write() int {
+	for req, filename := range w.requestToFile {
+		_, body, err := w.do(req)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
```

Comment:
All our existing `fmt.Println(...)`s have some context string prepended to them, like:

    fmt.Fprintln(os.Stderr, "error executing HTTP request:", err)

Would be good to have here (and below) too.